dali/core/mm/async_pool_test.cu

```diff
@@ -0,0 +1,325 @@
+        blk = blocks.back();
+        blocks.pop_back();
+      }
+      if (blk.stream != stream) {
+        if (stream) {
+          CUDA_CALL(cudaEventRecord(event, blk.stream));
+          CUDA_CALL(cudaStreamWaitEvent(stream, event, 0));
+        } else {
+          CUDA_CALL(cudaStreamSynchronize(blk.stream));
+        }
+      }
+      Check<<<div_ceil(blk.size, 1024), 1024, 0, stream>>>(
+            blk.ptr, blk.size, blk.fill, failure_buf);
+      if (stream) {
+        pool.deallocate_async(blk.ptr, blk.size, sv);
+      } else {
+        cudaStreamSynchronize(stream);
+        pool.deallocate(blk.ptr, blk.size);
+      }
+    }
+  }
+  copyD2H<int>(&failures, failure_buf, 1, stream);
+  cudaStreamSynchronize(stream);
+  ASSERT_EQ(failures, 0);
+}
+
+}  // namespace
+
+TEST(MMAsyncPool, SingleStreamRandom) {
+  CUDAStream stream = CUDAStream::Create(true);
+  test::test_device_resource upstream;
+
+  {
+    async_pool_base<memory_kind::device, free_tree, std::mutex> pool(&upstream);
+    vector<block> blocks;
+    detail::dummy_lock mtx;
+    AsyncPoolTest(pool, blocks, mtx, stream);
+  }
+
+  CUDA_CALL(cudaStreamSynchronize(stream));
+  std::cerr << "Peak consumption:     " << upstream.get_peak_size() << " bytes\n";
+  std::cerr << "Upstream allocations: " << upstream.get_num_allocs() << std::endl;
+  upstream.check_leaks();
+}
+
+TEST(MMAsyncPool, MultiThreadedSingleStreamRandom) {
+  CUDAStream stream = CUDAStream::Create(true);
+  mm::test::test_device_resource upstream;
+  {
+    vector<block> blocks;
+    std::mutex mtx;
+
+    async_pool_base<memory_kind::device, free_tree, std::mutex> pool(&upstream);
+
+    vector<std::thread> threads;
+
+    for (int t = 0; t < 10; t++) {
+      threads.push_back(std::thread([&]() {
+        AsyncPoolTest(pool, blocks, mtx, stream);
+      }));
+    }
+    for (auto &t : threads)
+      t.join();
+  }
+  CUDA_CALL(cudaStreamSynchronize(stream));
+  std::cerr << "Peak consumption:     " << upstream.get_peak_size() << " bytes\n";
+  std::cerr << "Upstream allocations: " << upstream.get_num_allocs() << std::endl;
+  upstream.check_leaks();
+}
+
+TEST(MMAsyncPool, MultiThreadedMultiStreamRandom) {
+  mm::test::test_device_resource upstream;
+  {
+    async_pool_base<memory_kind::device, free_tree, std::mutex> pool(&upstream);
+
+    vector<std::thread> threads;
+
+    for (int t = 0; t < 10; t++) {
+      threads.push_back(std::thread([&]() {
+        CUDAStream stream = CUDAStream::Create(true);
+        vector<block> blocks;
+        detail::dummy_lock mtx;
+        AsyncPoolTest(pool, blocks, mtx, stream);
+        CUDA_CALL(cudaStreamSynchronize(stream));
+      }));
+    }
+    for (auto &t : threads)
+      t.join();
+  }
+  std::cerr << "Peak consumption:     " << upstream.get_peak_size() << " bytes\n";
+  std::cerr << "Upstream allocations: " << upstream.get_num_allocs() << std::endl;
+  upstream.check_leaks();
+}
+
+
+TEST(MMAsyncPool, TwoStream) {
```

Comment:
I would move this to the end as it doesn't use `AsyncPoolTest` while one before and after use it.

Reply:
I moved it to the front.